[WIP] NL-means denoising

emmanuelle opened the pull request:
Following a message on the mailing-list, here is a first attempt at contributing my non-local means denoising code. It still lacks an example for the gallery. I would also welcome suggestions to improve the speed, since this algorithm is very slow...

emmanuelle commented on skimage/restoration/_nl_means_denoising.pyx:
```diff
@@ -0,0 +1,385 @@
+    cdef DTYPE_t [:, :, ::1] padded = np.ascontiguousarray(util.pad(
+                                        image.astype(np.float32),
+                                        offset, mode='reflect'))
+    cdef DTYPE_t [:, :, ::1] result = padded.copy()
```
hum, not sure how this could be tested...

emmanuelle commented:
I have now finished the implementation of both the "regular" non-local means for 2D, 2D+RGB and 3D, and of the fast method for 2D. Since the result is not exactly the same, I think we should keep the two methods - at the risk of cluttering up the API, so I'm interested in other opinions about this point : shall we keep only the fast one (which is slightly less precise than the slow one, according to http://www.ipol.im/pub/art/2014/120/article.pdf, but not very much so) or both? 

Also, for the sake of memory saving I am now using float32 data type: for my application (very large tomography datasets) the factor of 2 really makes a difference, but most scikit-image functions work with double precision. (as a result, I can't use `img_as_float` here, for example). Shall we have a `double_precision` kw argument set to True by default, but that can be toggled to False to reduce the memory cost?

@josephsalmon @jni @stefanv

jni commented:
@emmanuelle 

- re methods, I'd slightly favour keeping both. The speedup is massive, but in some applications (connectomics) a small increase in accuracy can save many hours of manual labour. Also, I think it's good to have reference implementations of key algorithms, in addition to any improvements we might want to add.

- re float32, where do you think things will fail? A single precision image should still work out of the box with most functions in skimage? I think a `double_precision` kwarg is a bit inelegant...

JDWarner commented:
I'd say default to the (much) faster mode which will work for most people in most applications. Allow those who want full precision to get it by asking.

jni commented:
@JDWarner :+1:!

josephsalmon commented:
@emmanuelle : I would keep the fast version. It would be more prudent to check the "the lack of precision" one would get. But I believe this would be way smaller than the noise contribution in the image. If not, one should not denoise such an image in the first place...

emmanuelle commented:
Thank you for all the answers! 

@jni I don't think float32 will result in a significant lack of precision. My concern was that we seldom use float32 in scikit-image.

@JDWarner good idea, maybe I'll change the names of the functions to be sure that people use the fast algorithm by default. Or we could have only one function in the API with a keyword argument for the algorithm used.

@josephsalmon indeed :-). I started writing an example for the gallery (https://github.com/emmanuelle/scikits.image/blob/nlm_denoise/doc/examples/plot_nonlocal_means.py), and the result is a bit counterintuitive. The peak signal to noise ratio of the noisy image (as compared to the original lena image) is slightly higher than the PSNR of the denoised image. I guess the reason is that the lena image is already corrupted by some noise, so that it is not the "true reference" for the non-local means. I had started comparing only the PSNR of outputs of the two versions of non-local means, but it is indeed interesting to compute also the PSNR of the noisy image!

jni commented:
@emmanuelle I didn't think that there would be a precision issue, rather, perhaps we have some hardcoded Cython functions that depend on a float64 input. However, I believe that where those exist, the image is first converted with `img_as_float`, so really, there should be no issue at all with using `float32`.

As to the API, I do suggest that you use a single function, with a kwarg to select between the two methods, and that defaults to the fast method.

Finally, regarding the gallery example, we are phasing out use of `lena`. Perhaps use `astronaut`, which is probably a better reference image anyway? =)

JDWarner commented:
@jni the Cython point is valid. However, it would be trivial to work in float32 then convert the result back to `np.float64`, which will never be lossy. Noted well in the docstring that shouldn't be a problem. The bigger issue is potential precision loss when coercing the input to float32.

emmanuelle commented:
New features: fast algorithm is now implemented for 3D, 2D and 2D-RGB. @josephsalmon it took me a while to figure out the recursion for 3-D images, but now it works!! @jni I hope you'll appreciate the 3D version :-D

I have also changed the API so that we expose only one function ``nl_means_denoising`` to the user, with a ``fast_mode`` keyword argument to choose the algorithm.

I think I have now implemented all that I wanted to put in this module, so I would welcome a new round of reviewing.

jni commented on skimage/restoration/_nl_means_denoising.pyx:
```diff
@@ -0,0 +1,596 @@
+cimport cython
+from libc.math cimport exp
+
+ctypedef np.float32_t DTYPE_t
```
I haven't seen this pattern before... Is it here in case we later want to change the dtype to float64? If so, could we use a different alias from `DTYPE_t`? It's not very descriptive. I'm thinking something like `IMGDTYPE`.

emmanuelle replied:
Well, this was an attempt to imitate the cython tutorial http://docs.cython.org/src/tutorial/numpy.html#adding-types 
I can indeed change the name of the alias.

jni replied:
... Well? =P

jni commented on skimage/restoration/_nl_means_denoising.pyx:
```diff
@@ -0,0 +1,596 @@
+@cython.boundscheck(False)
+cdef inline float patch_distance_2d(DTYPE_t [:, :] p1,
+                                    DTYPE_t [:, :] p2,
+                                    DTYPE_t [:, ::] w, int s):
```
Could you add a numpy-style docstring to this? (you can refer to this one in the other similar functions) I can't stand reading non-docstring-ed functions anymore. I want to know what the function does and what each of the inputs is, especially with single-char names like `w` and `s`.

emmanuelle replied:
ok, will do

jni replied:
These three patch distance functions can be replaced by a single nD function:

```python
@cython.boundscheck(False)
cdef inline float patch_distance_r(DTYPE_t [::1] p1,
                                   DTYPE_t [::1] p2,
                                   DTYPE_t [::1] w, int s):
    cdef int i
    cdef int center = s / 2
    cdef float tmp_diff = p1[center] - p2[center]
    cdef float init = w[center] * tmp_diff * tmp_diff
    if init > 1:
        return 0.
    cdef float distance = 0
    for i in range(s):
        if distance > 5:
            return 0.
        tmp_diff = p1[i] - p2[i]
        distance += w[i] * tmp_diff * tmp_diff
    return exp(-distance)
```

Then instead of passing in `patch[i0:i1, j0:j1]`, you pass `patch[i0:i1, j0:j1].ravel()`, and make sure s is the square of the current definition, ie for patch size 7 you pass in 49, etc.

I thought that this might adversely affect performance, but it seems to have almost no effect. Before:

```python
In [5]: %timeit denoise = nl_means_denoising(noisy, 7, 9, 0.08)
1 loops, best of 3: 3.25 s per loop
```

after:

```python
In [5]: %timeit denoise = nl_means_denoising(noisy, 7, 9, 0.08)
1 loops, best of 3: 3.27 s per loop
```

jni commented on skimage/restoration/_nl_means_denoising.pyx:
```diff
@@ -0,0 +1,596 @@
+    cdef float distance = 0
+    for i in range(s):
+        # exp of large negative numbers will be 0, so we'd better stop
+        if distance > 5:
```
I would suggest (1) replacing "5" with a module-scope "constant", `max_distance`, and (2) matching the value of said `max_distance` to `eps`, so that the breakpoint corresponds to minimum value of the distance. ie either increase `eps` to `exp(-5)` or decrease `max_distance` to `-log(1e-8)` (about 18). Does this make sense?

emmanuelle replied:
Thanks for noticing this inconsistency. In fact, setting ``eps`` to a non-zero value was just a trick to avoid divisions by zero, but I think there is no risk of division by zero. We could then just return 0. The idea of such a cut-off for distance is that is the two patches are sufficiently different from each other, then there is no good reason to average the central pixels together. And this condition saves in fact a lot of computing time.

jni replied:
I could see how it would be very good for compute time. Setting to 0 works, as it makes it more readable and doesn't feel like a "special" number.

jni commented:
@emmanuelle I'm *very* pleased with 3D support, but can you time the 2D function vs the 3D function with a 2D image with leading dimension of size 1? There is a lot of code redundancy... It would be good if those two functions could be merged...

jni commented:
@emmanuelle and similarly for RGB vs color... If there are big performance gains, the code redundancy makes sense, but otherwise it would be great to have just a single function...

jni commented on skimage/restoration/_nl_means_denoising.pyx:
```diff
@@ -0,0 +1,596 @@
+            for k in range(s):
+                tmp_diff = p1[i, j, k] - p2[i, j, k]
+                distance += w[i, j, k] * tmp_diff * tmp_diff
+    distance = exp(- distance)
```
unary operators should not be separated by a space from their argument: `-distance`

jni commented on skimage/restoration/_nl_means_denoising.pyx:
```diff
@@ -0,0 +1,596 @@
+    ----------
+    image: ndarray
+        input data to be denoised
+
```
Unnecessary blank lines in this docstring

jni commented on skimage/restoration/_nl_means_denoising.pyx:
```diff
@@ -0,0 +1,596 @@
+    if s % 2 == 0:
+        s += 1  # odd value for symmetric patch
+    cdef int n_x, n_y
+    n_x, n_y = image.shape
```
I hate to bring this up with so much invested already, but... `n_r, n_c`? ;)

emmanuelle replied:
no worries :-). When I recently read your PR #1280 about coordinates it occurred to me that I should change my naming conventions, but I was too lazy to do it. That's one of the beauties of code review that laziness becomes less acceptable :-)

One thing, though, is that I'm not completely convinced by the name "frame" in 3D. How about "slice"? I don't know if you had the time to read my recent comments in #1280.

jni replied:
`slice` is a Python keyword. =\ My vote was for `level`, but people didn't like the 1 letter abbreviation (`l` or `v`, both have problems). Although I agree with you in not liking frame, it was the compromise acceptable to most people.

emmanuelle replied:
oh yes indeed, I must be a bit tired :-).

jni commented on skimage/restoration/_nl_means_denoising.pyx:
```diff
@@ -0,0 +1,596 @@
+    cdef float new_value
+    cdef float weight_sum, weight
+    xg, yg, zg = np.mgrid[-offset: offset + 1, -offset: offset + 1,
+                            -offset: offset + 1]
```
PEP8 prescribes: `[-offset : offset+1, -offset : offset+1, -offset : offset+1]`

I'd prefer: `[(slice(offset, offset+1),) * image.ndim]` ;)

jni commented on skimage/restoration/_nl_means_denoising.pyx:
```diff
@@ -0,0 +1,596 @@
+    xg, yg, zg = np.mgrid[-offset: offset + 1, -offset: offset + 1,
+                            -offset: offset + 1]
+    cdef DTYPE_t [:, :, ::1] w = np.ascontiguousarray(np.exp(
+                            - (xg ** 2 + yg ** 2 + zg ** 2) / (2 * A ** 2)).
```
`-(fg**2 + rg**2 + cg**2) / (2 * A**2)`

jni commented on skimage/restoration/_nl_means_denoising.pyx:
```diff
@@ -0,0 +1,596 @@
+    cdef int x_start_i, x_end_i, y_start_j, y_end_j, z_start_k, z_end_k
+    w = 1. / (np.sum(w) * h ** 2) * w
+    # Coordinates of central pixel and patch bounds
+    for x in range(offset, n_x + offset):
```
This *might* be doable with `np.nditer`. I don't know how much worse that would be for performance, but it would certainly help readability!

jni commented on skimage/restoration/_nl_means_denoising.pyx:
```diff
@@ -0,0 +1,596 @@
+            else:
+                alpha = 1.
+            integral = np.zeros_like(padded)
+            for x in range(max(1, - t1), min(n_x, n_x - t1)):
```
A lot of the functions in this module are really huge... I think it would help a lot if some of these loops became their own inline functions. For example, here you could replace this block with `integral = _compute_patch_integral(padded, n_r, t1, n_c, t2)`

jni commented on skimage/restoration/_nl_means_denoising.pyx:
```diff
@@ -0,0 +1,596 @@
+                    result[x + t1, y + t2] += weight * padded[x, y]
+    for x in range(offset, n_x - offset):
+        for y in range(offset, n_y - offset):
+            # I think there is no risk of division by zero
```
Haha I love this comment. It's like, "This is tooootally probably legit." =P

emmanuelle replied:
Well, maybe I should remove "I think" then. Since there is always a weight of one for the (0, 0) shift, there is no risk of division by zero :-).

jni replied:
yeah that's a good idea. =)

jni commented on skimage/restoration/non_local_means.py:
```diff
@@ -0,0 +1,125 @@
+    ----------
+    image : ndarray
+        input data to be denoised
+
```
Remove these blank lines (I think)

jni commented on skimage/restoration/non_local_means.py:
```diff
@@ -0,0 +1,125 @@
+        else:
+            return np.array(_nl_means_denoising_3d(image, patch_size,
+                                patch_distance, h))
+    if image.ndim == 3 and image.shape[-1] == 3:  # 2-D color (RGB) images
```
I'm quite happy with this, but when we discussed this "magic" in the context of my SLIC PR, the conclusion was that it's better to have a user flag, such as `rgb=True`, that determines whether an image is interpreted as 3D or 2D+c. The rationale was provided by @JDWarner: if someone has a little test 3D image with, say, `image.shape[-1] == 5`, on which they are developing their whole code, they will do the NLmeans denoising and be happy with their code. Then when they deploy it on many test images, they might come across a 3D image with `image.shape[-1] == 3`, and the code will do something completely different. In the case of a kwarg, users are protected from such a scenario.

The flag name is `multichannel` for SLIC and although it only makes much sense with RGB here, I would suggest we maintain the same flag name across all functions.

Also, right now, 3D images with exactly 4 columns in the last dimension just fail. =P

emmanuelle replied:
OK, I think both options (using "magics" or no using it) have their pros and cons; but since you already had the discussion about SLIC I'm happy to be consistent with existing functions. I'll add a ``multichannel`` flag then.

jni commented on skimage/restoration/non_local_means.py:
```diff
@@ -0,0 +1,125 @@
+            return np.array(_nl_means_denoising_2drgb(image, patch_size,
+                                patch_distance, h))
+    else:
+        raise ValueError("Non local means denoising is only possible for \
```
I would say this is a `NotImplementedError`: "NLM is only *implemented* for 2D grayscale, 2D RGB, and 3D grayscale images."

emmanuelle replied:
ok!

emmanuelle commented:
@jni about code redundancy and whether 2D images could be processed with the 3D function, the answer is unfortunately no. I'm not even sure whether this could work with non-square / non-cubic patches, which we don't support in this code.

But I think that I should be able to merge the 2D and 2D RGB functions. Most of the code is the same, including the loops over patches. It's just that I don't know how to write in a dimension-independent way lines such as ``cdef DTYPE_t [:, :, ::1] result = np.zeros_like(padded)``. Of course, I can use flags such as ``multichannel == True`` but if you know other tricks for such situations, please let me know!!

emmanuelle commented:
@jni thanks a lot for the careful reviewing. I have a couple of questions for you, see my answers. Now back to the code to implement your suggestions!

jni commented:
@emmanuelle to merge RGB/single value code, why not add something to the Python wrapper like:

```python
if not multichannel:
    image = image[..., np.newaxis]
```

Then always use the RGB code? (just making sure not to hardcode the number of channels).

Regarding 2D/3D code compatibility, did you ever see my [blog post](http://ilovesymposia.com/2014/06/24/a-clever-use-of-scipys-ndimage-generic_filter-for-n-dimensional-image-processing/) about @vighneshbirodkar's use of `numpy.nditer`? I think you might be able to *at least* get rid of the deeply nested loops. Maybe. Just a thought. =)

emmanuelle commented:
@jni thanks for the link, I must confess I didn't know your blog! I have been away from scikit-image development for too long :-).

I can indeed add a dimension to 2D grayscale images. It's just that in the Cython code, when initializing an array with ``cdef DTYPE_t [:, ::1] result = np.zeros_like(padded)`` I understand that it is performance-critical to specify that the array is C-contiguous and that elements along the last dimension are one element apart in memory (which is not very interesting if the last dimension has only one element). But maybe this notation just means that the array is C-contiguous and the performance will be the same on an array of shape (n_x, n_y) and (n_x, n_y, 1). Checking a possible performance change is the easiest way to be sure of that!

jni commented:
@emmanuelle you were too busy being interviewed for science magazines! Totally understandable! =P

Not to mention Caribbean cruises. =P

`::1` merely indicates C-contiguity, which is true regardless of an extra dimension. The performance should be very close. But, as you say, measurement is better than theory! =)

emmanuelle commented:
@jni oh yeah life is so hard :-). 

I think I have now addressed all your comment, thanks a ton for your suggestions. I think it's getting close to being mergeable. What do you think?

jni commented on skimage/restoration/_nl_means_denoising.pyx:
```diff
@@ -0,0 +1,565 @@
+    p1 : 2-D array_like
+        first patch
+    p2 : 2-D array_like
+        first patch
```
second patch, surely. =P

jni commented on skimage/restoration/_nl_means_denoising.pyx:
```diff
@@ -0,0 +1,565 @@
+    p2 : 2-D array_like
+        first patch
+    w : 2-D array_like
+        array of weigths for the different pixels of the patches
```
typo: weights

stefanv replied:
array -> Array, fullstop at the end

stefanv commented on skimage/restoration/_nl_means_denoising.pyx:
```diff
@@ -0,0 +1,565 @@
+    ----------
+    p1 : 3-D array_like
+        first patch, 2D image with last dimension corresponding to channels
+    p2 : 3-D array_like
```
f -> first, full stop etc.

jni commented on skimage/restoration/_nl_means_denoising.pyx:
```diff
@@ -0,0 +1,565 @@
+        for j in range(s):
+            tmp_diff = p1[i, j] - p2[i, j]
+            distance += (w[i, j] * tmp_diff * tmp_diff)
+    distance = exp(- distance)
```
`-distance` (no space)

jni commented on skimage/restoration/_nl_means_denoising.pyx:
```diff
@@ -0,0 +1,565 @@
+
+    Parameters
+    ----------
+    image: ndarray
```
shape (M, N, C)

jni commented on skimage/restoration/_nl_means_denoising.pyx:
```diff
@@ -0,0 +1,565 @@
+    cdef float tmp_diff
+    for i in range(s):
+        # exp of large negative numbers will be 0, so we'd better stop
+        if distance > 5:
```
can you define a constant at the top, e.g. `MAXD`, `MAX_DISTANCE`, or `DISTANCE_CUTOFF`, and use it here? This will make the code a bit more readable.

jni replied:
Same for other places where this cutoff appears, of course.

stefanv commented on skimage/restoration/_nl_means_denoising.pyx:
```diff
@@ -0,0 +1,565 @@
+
+    Notes
+    -----
+    The returned distance is given by
```
Invalid math markup

stefanv commented on skimage/restoration/_nl_means_denoising.pyx:
```diff
@@ -0,0 +1,565 @@
+                    int x_pln, int x_row, int x_col, int offset,
+                    float s_cube_h_square):
+    """
+    See
```
References section

jni commented on skimage/restoration/_nl_means_denoising.pyx:
```diff
@@ -0,0 +1,565 @@
+    h: float, optional
+        cut-off distance (in gray levels). The higher h, the more permissive
+        one is in accepting patches.
+    """
```
Returns?

jni commented on skimage/restoration/_nl_means_denoising.pyx:
```diff
@@ -0,0 +1,565 @@
+    w = 1. / (np.sum(w) * h ** 2) * w
+    # Coordinates of central pixel and patch bounds
+    for x_pln in range(offset, n_pln + offset):
+        x_pln_start = x_pln - offset
```
Why `x_foo`? I think `for pln in range(offset, pln + offset):`, `for row in range(offset, n_row + offset)`, etc, would work fine.

jni commented on skimage/restoration/_nl_means_denoising.pyx:
```diff
@@ -0,0 +1,565 @@
+                x_col_end = x_col + offset + 1
+                new_value = 0
+                weight_sum = 0
+                # Coordinates of test pixel and patch bounds
```
Could we use `skimage.util.view_as_windows` here to simplify the logic, and `np.nditer` to reduce the level of nesting?

```python
paddedw = view_as_windows(padded, (s,) * 3)
iterator = np.nditer(image, ["multi_index"])
while not iterator.finished:
    pln, row, col = [x + offset for x in iterator.multi_index]
    central_patch = paddedw[pln, row, col]
```

There's probably a similar reduction to be had with the inner iteration... What I don't know is how much of a performance hit this is.

jni replied:
It [looks like](http://docs.scipy.org/doc/numpy/reference/arrays.nditer.html#putting-the-inner-loop-in-cython) `np.nditer` can be quite performant in Cython! (when provided with dtype info, which we can do.)

jni commented on skimage/restoration/_nl_means_denoising.pyx:
```diff
@@ -0,0 +1,565 @@
+                    alpha = 0.5
+                else:
+                    alpha = 1.
+                integral = np.zeros_like(padded)
```
Is this the same as `skimage.transform.integral_image`?

emmanuelle replied:
yes, integral turns out to be the same array as 

```
integral_image((padded[t_row:, t_col, :] - padded[:-t_row, :-t_col])**2)
```

Thanks for noticing this, I was implementing the formula provided in the article of Froment to compute ``integral``, and then I wasn't thinking much about it. 

It will remove a lot of code that is hard to read if I use ``integral_image``. However, it would also use more memory if ``(padded[t_row:, t_col, :] - padded[:-t_row, :-t_col])**2`` is computed -- which is not a problem for most images, but might be a problem for large 3D images. Well, I can inline a custom integral function with this part of the code, and write a test to check whether the result is the same as for ``ìntegral_image``. Unless you have a better idea...

jni commented on skimage/restoration/_nl_means_denoising.pyx:
```diff
@@ -0,0 +1,565 @@
+                # I think there is no risk of division by zero
+                # except in padded zone
+                result[x_pln, x_row, x_col] /= weights[x_pln, x_row, x_col]
+    return result[pad_size: - pad_size, pad_size: - pad_size,
```
Please be consistent with unary `-` (no space between `-` and operand, ie, `-pad_size`). Also, PEP8 prescribes either space on either side of `:`, or no space. Always the same. ie: `pad_size:(-pad_size)` or `pad_size : -pad_size`.

jni commented on skimage/restoration/non_local_means.py:
```diff
@@ -0,0 +1,125 @@
+    """
+    if image.ndim == 2:
+        image = image[..., np.newaxis]
+        if fast_mode:
```
These are the same as below, so you don't need them:

```python
if image.ndim == 2:
    image = image[..., np.newaxis]
if image.ndim != 3:
    raise NotImplementedError
if multichannel:
    # [...]
    # squeeze here anyway, it is a no-op for non-singleton dimensions
else:
    # [...]
```

jni replied:
Oh, I just realised you need one more thing:
```
if image.ndim == 2:
    image = image[..., np.newaxis]
    multichannel = True
```

jni commented on skimage/restoration/_nl_means_denoising.pyx:
```diff
@@ -0,0 +1,565 @@
+                                    padded[x_row + t_row, x_col + t_col, 0])**2
+                    else:
+                        distance = ((padded[x_row, x_col, 0] -
+                                    padded[x_row + t_row, x_col + t_col, 0])**2
```
align with correct opening brace

jni commented on skimage/restoration/_nl_means_denoising.pyx:
```diff
@@ -0,0 +1,565 @@
+                    else:
+                        distance = ((padded[x_row, x_col, 0] -
+                                    padded[x_row + t_row, x_col + t_col, 0])**2
+                                    +(padded[x_row, x_col, 1] -
```
binary operators go at the end of the last line

jni commented:
@emmanuelle there's still a few improvements to be made... I'm actually reluctant to merge this as long as the six-fold nested loops remain... I'm positive that they can be effectively flattened, but I don't have time right now to spend on doing so myself... I've provided a few clues if you want to tackle it, otherwise I definitely will at some point...!

emmanuelle commented:
@jni thanks a lot for your patience :-)

Thanks to your comments and the one by Stéfan, I've made quite a few improvements. I think the code is now clearer, with new inline helper functions, clearer variable names, and some additional comments. 

However, I have to say that I'm not convinced by the idea of using ``np.nditer``. In particular, I don't think it would be an improvement in terms of code readability, with many flags to provide, and some additional syntax. Plus, the use of helper functions becomes more complicated. I think it's no wonder if ``scikit-image`` does not use ``np.nditer`` at all in its codebase, and neither does ``scikit-learn`` or ``scipy``. 

Of course, the nested loops are not great, but I've made sure to add comments so that when the reader sees that one is looping over plane, column and row, she might consider it a "single" loop on pixel coordinates, without too much effort.

Tell me what you think, and if you keep thinking that ``np.nditer`` is a better idea, maybe we can ask for another advice? Thanks a lot anyway!!

ahojnnes commented:
@emmanuelle This is impressive work! I personally prefer the for loops, even though it is probably not as generic in terms of N-D processing...

jni commented:
@emmanuelle @ahojnnes I don't have anything about nested for loops per se, but when the nesting runs six levels deep, I cringe. We don't use `nditer` but I think @stefanv's [`ndindex` wrapper](https://github.com/nipy/dipy/blob/master/dipy/core/ndindex.py) has some very nice semantics: `for pln, row, col in ndindex(arr.shape):`. Doesn't get any cleaner than that! Another option is `itertools.product`, which further allows incomplete ranges (something `ndindex` doesn't yet do):

`for pln, row, col in product(range(offset, n_planes + offset), range(offset, n_rows + offset), range(offset, n_cols + offset)):`

jni commented:
Incidentally, don't take my lack of use of this as evidence that it's not useful or clear — I was simply not aware of it until last summer! ;)

JDWarner commented:
I wasn't aware `itertools.product` could be used in that way. Is it Cython-friendly and performant?

jni commented:
@JDWarner that's the million dollar question! I don't actually know how it would compare. I contend that even a 2x slowdown would be worthwhile to improve code clarity, though. (But I'd concede if it exceeded that.)

emmanuelle commented:
@jni well, I learnt the existence of ``np.nditer`` thanks to your suggestions, so I won't blame you for discovering it last summer :-)

For me, 3 nested loops on the 3 axes of an image array are technically 3 nested ``for`` loops, but semantically they can be grasped as one big loop over the coordinates of an array, so I don't consider it to be that bad.

I totally agree that it would be much better to have something like

```
for pln, row, col in iterate_coordinates(images):
     do_something()
```

and this is what ``np.nditer`` is trying to do, but the syntax is not so nice either (-yet : it might evolve).

emmanuelle commented:
So, I gave it a try with ``product``, and it turns out that the execution time is 40 % larger with ``product`` for a 512x512 image, and 2x longer for a 60x60x60 3-D image.

jni commented:
Hi @emmanuelle! Thank you very much for trying those benchmarks! It's amusing that it's exactly on my threshold for when it's not worth doing. =P

I'll let @stefanv be the judge of whether he's satisfied with the code as it stands (it's certainly appealing to have NL-means in 0.11!). But I would argue that it's a *major*, not slight, improvement to readability, and when we merge I would add an issue tagged with 0.12 for improving the iteration structure. I think a general nd iterator could be made in Cython for this purpose and could actually be used in many parts of the library (e.g. SLIC)!

stefanv commented:
@emmanuelle Can you paste the other code for comparison, please?

JDWarner commented:
@jni Frankly I would disagree that nested loops hurts readability. To me, aggressively vectorized functions and other clever methods usually end up being much harder to follow. I don't have a problem with them, but just doubt readability is a valid criticism.

With a little tweaking of the whitespace and comments, I find it hard to believe this could be easier to follow/maintain with any other algorithmic structure. The added bonus is that looping in this manner generally is how papers describe algorithms, so it's easier to compare and search for potential discrepancies vs. the original publications.

``` python 
# Iterate over rows, taking padding into account
for row in range(offset, n_row + offset):
    row_start = row - offset
    row_end = row + offset + 1
    
    # Iterate over columns, taking padding into account
    for col in range(offset, n_col + offset):
        col_start = col - offset
        col_end = col + offset + 1
    	
    	# Iterate over channels
        for color in range(n_ch):
            new_values[color] = 0
            
        # Reset weights for each local region
        weight_sum = 0
        
        # Iterate over local 2d patch for each pixel
        # First rows
        for i in range(max(-d, offset - row),
                       min(d + 1, n_row + offset - row)):
            row_start_i = row_start + i
            row_end_i = row_end + i
            
            # Local patch columns
            for j in range(max(-d, offset - col),
                           min(d + 1, n_col + offset - col)):
                col_start_j = col_start + j
                col_end_j = col_end + j
                
                # Shortcut for grayscale, else assume RGB
                if n_ch == 1:
                    weight = patch_distance_2d(
                             padded[row_start:row_end,
                                    col_start:col_end, 0],
                             padded[row_start_i:row_end_i,
                                    col_start_j:col_end_j, 0],
                             w, s)
                else:
                    weight = patch_distance_2drgb(
                             padded[row_start:row_end,
                                    col_start:col_end, :],
                             padded[row_start_i:row_end_i,
                                    col_start_j:col_end_j, :],
                                    w, s)
                
                # Collect results in weight sum
                weight_sum += weight
                
                # Apply to each channel multiplicatively
                for color in range(n_ch):
                    new_values[color] += weight * padded[row + i,
                                                         col + j, color]
                                                         
        # Normalize the result
        for color in range(n_ch):
            result[row, col, color] = new_values[color] / weight_sum
            
# Return cropped result, undoing padding
return result[offset:-offset, offset:-offset]
```

ahojnnes commented:
@JDWarner As already stated above, I agree with your argument.

jni commented:
Ok, now show me the same code for 3D + ch and then we'll talk. ;)

With 6-fold nesting, by the time you're in the innermost loop, it's really hard to keep in mind what all the variables are (e.g. which one is the inner iteration), and you've probably scrolled past the place where they were originally defined. I think the rule of thumb that functions should not be longer than 25 lines is *very* good and should be more strictly followed. It is, however, impossible to follow for functions with such deep nesting.

JDWarner commented:
I've got some 3d + multichannel Cython done with loops. The issue with readability isn't so much keeping track of stuff - that should be fine, if informative variable names are used - as much as the 80 char line requirement constraining the key innermost statements. Deliberately violating PEP8 by either cutting indentation spacing in half or allowing slightly longer lines largely resolves these issues.

Also, and perhaps most importantly, by structuring in this way we're inherently getting things as ready as possible for parallelization. The entire function is factored into its minimal form/kernel. If/when Cython supports OpenMP through LLVM on newer variants of OSX, the big loop will fly with the addition of `with nogil:` and `prange`. I use this for several personal functions, and it's awesome. This structure is also likely superior for Numba/Theano/etc. gains. If fancier methods are used and then parallelization question arises, we'd probably end up retracing our steps.

My general approach is to use fancier approaches for more limited things, like operations per-pixel, but when doing anything involving local patches naive loops w/parallelization is the clear path forward.

emmanuelle commented:
thank you very much guys for your comments! I'll add some comments as @JDWarner suggested.

@stefanv For comparison, the code using ``product`` looks like

 ```
# Outer loops omitted...
                # Inner loops on pixel coordinates
                for pln in range(pln_dist_min, pln_dist_max):
                     for row in range(row_dist_min, row_dist_max):
                          for col in range(col_dist_min, col_dist_max):
                               distance = _integral_to_distance_3d(integral,
                                        pln, row, col, offset, s_cube_h_square)
                               ....
```

to be compared to

```
# Outer loops omitted...
                # Inner loops on pixel coordinates
                for pln, row, col in product(range(pln_dist_min, pln_dist_max),
                                                         range(row_dist_min, row_dist_max),
                                                          range(col_dist_min, col_dist_max)):
                     distance = _integral_to_distance_3d(integral,
                                        pln, row, col, offset, s_cube_h_square)

```

emmanuelle commented:
Would you think that this PR is mergeable now?

JDWarner commented on skimage/restoration/_nl_means_denoising.pyx:
```diff
@@ -0,0 +1,712 @@
+    Used in _fast_nl_means_denoising_2d
+    """
+    cdef float distance
+    distance =  integral[row + offset, col + offset] + \
```
Would precalculating these offsets save half the addition/subtraction operations? Or is Cython smart enough to do this without being told? Might be worthwhile, as it's the innermost loop.

``` python
row_pos = row + offset
row_neg = row - offset
col_pos = col + offset
col_neg = col - offset
```

Doing this for 3d in the next function actually cuts out 75% of the simple math ops.

emmanuelle commented:
@JDWarner I gave it a try (I had alreary tried such tricks before moving parts of the code to helper functions, but I did it again for the computations you mention, computing the values of ``row_pos`` etc. as early as possible), and it turns out that it makes no difference at all...

ahojnnes commented:
+1 For merging this now, unless @emmanuelle wants to continue working on it? This is a great contribution as it stands now!

stefanv commented:
+1

JDWarner commented:
@emmanuelle apologies for making you go through that. This looks good to me too, so I'm hitting the green button!

blink1073 commented:
Woohoo!  That was a bear.  Well done @emmanuelle!

stefanv commented:
Great addition for the imminent release!
On Feb 19, 2015 5:20 PM, "Steven Silvester" <notifications@github.com>
wrote:

> Woohoo! That was a bear. Well done @emmanuelle
> <https://github.com/emmanuelle>!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-image/scikit-image/pull/874#issuecomment-75173809>
> .
>

emmanuelle commented:
@JDWarner no worries at all, I'm actually much grateful to y'all guys for making so many useful comments, and for all your patience during the numerous iterations :-). That said, I'm also quite glad that this PR is merged now ;-)! Thanks a lot!!
